File: dungeons-1.21.11/src/main/java/fishmod/utils/data/ItemUtil.java
package fishmod.utils.data;

import net.minecraft.client.MinecraftClient;
import net.minecraft.client.network.ClientPlayerEntity;
import net.minecraft.component.DataComponentTypes;
import net.minecraft.component.type.LoreComponent;
import net.minecraft.component.type.NbtComponent;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NbtCompound;
import net.minecraft.text.Text;

import java.util.List;

public class ItemUtil {

    public static String getId(ItemStack item) {
        NbtComponent nbt = item.get(DataComponentTypes.CUSTOM_DATA);
        if (nbt == null) return null;

        NbtCompound compound = nbt.copyNbt();

        return compound.getString("id", null);
    }

    /** Reads a raw string value out of an item's CUSTOM_DATA NBT (e.g. "petInfo"), or null. */
    public static String getNbtString(ItemStack item, String key) {
        NbtComponent nbt = item.get(DataComponentTypes.CUSTOM_DATA);
        if (nbt == null) return null;
        return nbt.copyNbt().getString(key, null);
    }

    public static String getUuid(ItemStack item) {
        NbtComponent nbt = item.get(DataComponentTypes.CUSTOM_DATA);
        if (nbt == null) return null;

        NbtCompound compound = nbt.copyNbt();

        return compound.getString("uuid", null);
    }

    public static boolean isHolding(String name) {
        ClientPlayerEntity player = MinecraftClient.getInstance().player;
        if (player == null) return false;
        return player.getMainHandStack().getName().getString().contains(name);
    }

    public static boolean itemHasName(ItemStack itemStack, String name) {
        ClientPlayerEntity player = MinecraftClient.getInstance().player;
        if (player == null || itemStack == null) return false;
        return itemStack.getName().getString().contains(name);
    }

    public static boolean containsLore(ItemStack item, String contain) {
        if (item ==null) return false;
        LoreComponent lore = item.get(DataComponentTypes.LORE);
        if (lore == null) return false;

        List<Text> lines = lore.lines();
        if (lines.isEmpty()) return false;

        for (Text line : lines.reversed()) {
            String string = line.getString();
            if (string.contains(contain)) {
                return true;
            }
        }
        return false;
    }

    public static boolean containsIgnoreCaseLore(ItemStack item, String contain) {
        if (item ==null) return false;
        LoreComponent lore = item.get(DataComponentTypes.LORE);
        if (lore == null) return false;

        List<Text> lines = lore.lines();
        if (lines.isEmpty()) return false;

        for (Text line : lines.reversed()) {
            String string = line.getString();
            if (string.toLowerCase().contains(contain.toLowerCase())) {
                return true;
            }
        }
        return false;
    }

    public static boolean containsNBT(ItemStack item, String contain) {
        NbtComponent nbt = item.get(DataComponentTypes.CUSTOM_DATA);
        if (nbt == null) {
            return false;
        }

        return  nbt.toString().contains(contain);
    }

}

File: dungeons-1.21.11/src/main/kotlin/fishmod/utils/data/ItemUtil.kt
package fishmod.utils.data

import net.minecraft.client.MinecraftClient
import net.minecraft.component.DataComponentTypes
import net.minecraft.item.ItemStack

object ItemUtil {

    @JvmStatic
    fun getId(item: ItemStack): String? {
        val nbt = item.get(DataComponentTypes.CUSTOM_DATA) ?: return null
        val compound = nbt.copyNbt()
        return compound.getString("id", null)
    }

    /** Reads a raw string value out of an item's CUSTOM_DATA NBT (e.g. "petInfo"), or null. */
    @JvmStatic
    fun getNbtString(item: ItemStack, key: String): String? {
        val nbt = item.get(DataComponentTypes.CUSTOM_DATA) ?: return null
        return nbt.copyNbt().getString(key, null)
    }

    @JvmStatic
    fun getUuid(item: ItemStack): String? {
        val nbt = item.get(DataComponentTypes.CUSTOM_DATA) ?: return null
        val compound = nbt.copyNbt()
        return compound.getString("uuid", null)
    }

    @JvmStatic
    fun isHolding(name: String): Boolean {
        val player = MinecraftClient.getInstance().player ?: return false
        return player.mainHandStack.name.string.contains(name)
    }

    @JvmStatic
    fun itemHasName(itemStack: ItemStack?, name: String): Boolean {
        val player = MinecraftClient.getInstance().player
        if (player == null || itemStack == null) return false
        return itemStack.name.string.contains(name)
    }

    @JvmStatic
    fun containsLore(item: ItemStack?, contain: String): Boolean {
        if (item == null) return false
        val lore = item.get(DataComponentTypes.LORE) ?: return false

        val lines = lore.lines()
        if (lines.isEmpty()) return false

        for (line in lines.asReversed()) {
            val string = line.string
            if (string.contains(contain)) {
                return true
            }
        }
        return false
    }

    @JvmStatic
    fun containsIgnoreCaseLore(item: ItemStack?, contain: String): Boolean {
        if (item == null) return false
        val lore = item.get(DataComponentTypes.LORE) ?: return false

        val lines = lore.lines()
        if (lines.isEmpty()) return false

        for (line in lines.asReversed()) {
            val string = line.string
            if (string.lowercase().contains(contain.lowercase())) {
                return true
            }
        }
        return false
    }

    @JvmStatic
    fun containsNBT(item: ItemStack, contain: String): Boolean {
        val nbt = item.get(DataComponentTypes.CUSTOM_DATA) ?: return false
        return nbt.toString().contains(contain)
    }
}
